crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs

```diff
@@ -185,3 +197,486 @@ pub(crate) fn scale_and_add_subtract_eval<S: Scalar>(
+    S: Scalar + From<L>,
+>(
+    lhs: &[L],
+    rhs: &[R],
+    alloc: &'a Bump,
+    is_right_bigger_int_type: bool,
+) -> (&'a [L], &'a [S]) {
+    let division_wrapped = alloc.alloc_slice_fill_with(lhs.len(), |_| L::zero());
+    let division = alloc.alloc_slice_fill_with(lhs.len(), |_| S::ZERO);
+    for (dw, d, &l, &r) in izip!(&mut *division_wrapped, &mut *division, lhs, rhs) {
+        *dw = if l == L::min_value() && r == -R::one() {
+            L::min_value()
+        } else if r == R::zero() {
+            L::zero()
+        } else if is_right_bigger_int_type {
+            NumCast::from(R::from(l).unwrap() / r).unwrap()
+        } else {
+            l / L::from(r).unwrap()
+        };
+        *d = S::from(*dw)
+            * (if *dw == L::min_value() && r == -R::one() {
+                -S::ONE
+            } else {
+                S::ONE
+            });
+    }
+    (division_wrapped, division)
+}
+
+/// Modulo two columns of data, where the data types are some unsigned int type(s).
+/// Note that `i128::MIN % -1`, for example, is unusual in that `i128::MIN / -1`
+/// ordinarily returns a value that is not containe dby i128. Division wraps this operation,
+/// but modulo still returns 0 here.
+/// Division by 0 returns the numerator for modulo.
+#[allow(clippy::missing_panics_doc)]
+fn modulo_integer_columns<
+    'a,
+    L: NumCast + Copy + PrimInt + Neg<Output = L>,
+    R: NumCast + Copy + PrimInt + Neg<Output = R>,
+    O: NumCast + PrimInt,
+>(
+    lhs: &[L],
+    rhs: &[R],
+    alloc: &'a Bump,
+    is_right_bigger_int_type: bool,
+) -> &'a [O] {
+    let remainder = alloc.alloc_slice_fill_with(lhs.len(), |_| O::zero());
+    remainder
+        .iter_mut()
+        .zip(lhs.iter().copied().zip(rhs.iter().copied()))
+        .for_each(|(m, (l, r))| {
+            *m = if l == L::min_value() && r == -R::one() {
+                O::zero()
+            } else if r == R::zero() {
+                NumCast::from(l).unwrap()
+            } else if is_right_bigger_int_type {
+                NumCast::from(R::from(l).unwrap() % r).unwrap()
+            } else {
+                NumCast::from(l % L::from(r).unwrap()).unwrap()
+            }
+        });
+    remainder
+}
+
+/// Divide one column by another. The first value in the tuple wraps `MIN / -1` back to `MIN`,
+/// whereas the second returns `-MIN`, where `MIN` is the minimum value of a signed int.
+/// For now, only signed integer types are supported.
+/// # Panics
+/// Panics if: `lhs` and `rhs` are not of the same length or column type division is unsupported.
+#[allow(clippy::too_many_lines)]
+#[cfg_attr(not(test), expect(dead_code))]
+pub(crate) fn divide_columns<'a, S: Scalar>(
+    lhs: &Column<'a, S>,
+    rhs: &Column<'a, S>,
+    alloc: &'a Bump,
+) -> (Column<'a, S>, &'a [S]) {
+    let lhs_len = lhs.len();
+    let rhs_len = rhs.len();
+    assert!(
+        lhs_len == rhs_len,
+        "lhs and rhs should have the same length"
+    );
```

Comment:
You should still test this.
https://doc.rust-lang.org/book/ch11-01-writing-tests.html#checking-for-panics-with-should_panic